zubbi/scraper/scraper.py

```diff
@@ -146,29 +159,44 @@ def get_file_info(self, path):
+        # https://pygithub.readthedocs.io/en/v2.1.1/examples/Repository.html#get-all-of-the-contents-of-the-repository-recursively
         try:
-            roles = self.repo.directory_contents(ROLES_DIRECTORY)
-            for role_name, remote_file in roles.items():
+            # Ansible requires roles to be defined in a certain directory structure.
+            # Thus, we are only interested in directories
+            dirs_to_search = [
+                item
+                for item in self.repo.directory_contents(ROLES_DIRECTORY).values()
+                if item.type == "dir"
+            ]
+            while dirs_to_search:
                 try:
-                    if remote_file.type != "dir":
-                        # Ansible requires the role to be defined in a
-                        # certain directory structure. Thus, we can
-                        # ignore it in case it's not a directory.
-                        continue
-                    last_changed = self.repo.last_changed(remote_file.path)
-                    existing_files = self.repo.directory_contents(remote_file.path)
-                    # Skip empty directories or files
-                    if not existing_files:
+                    dir = dirs_to_search.pop(0)
+                    dir_items = self.repo.directory_contents(dir.path)
+
+                    subdirs = [
+                        item for item in dir_items.values() if item.type == "dir"
+                    ]
+                    # When directory does not contain one of role mandatory directories
+                    # it is not a role and its subdirectories (if any) should be further
+                    # scanned.
+                    # This is done by appending those subdirectories to 'dirs_to_search'
+                    # search list (this implements the recursive search)
+                    if set([d.name for d in subdirs]).isdisjoint(ROLE_MANDATORY_DIRS):
+                        dirs_to_search.extend(subdirs)
                         continue
-                    readme_file = self.find_matching_file(README_FILES, existing_files)
-                    changelog_file = self.find_matching_file(
-                        CHANGELOG_FILES, existing_files
-                    )
-                    role_files[role_name] = {
+
+                    # Once the role is found, we are only interested in the timestamp of
+                    # the latest update (the last git change), README and CHANGELOG files
+                    # Those files should be on the top-level per role.
+                    last_changed = self.repo.last_changed(dir.path)
+                    readme_file = self.find_matching_file(README_FILES, dir_items)
+                    changelog_file = self.find_matching_file(CHANGELOG_FILES, dir_items)
+                    # role directory path (relative to ROLES_DIRECTORY) is the role name
+                    role_files[dir.path.replace(f"{ROLES_DIRECTORY}/", "", 1)] = {
```

Comment:
I think we should use Python's `pathlib` for this purpose:

```suggestion
                    role_files[Path(dir.path).relative_to(ROLES_DIRECTORY)] = {
```

That would need an additional import at the top of this file:
```python
from pathlib import Path
```

Reply:
hmm, is import really needed?
dir.path is already pathlib object. I will try to run the "relative_to" method without converting to Path first.
In the worst case I will import the lib and convert dir.path

Reply:
Sorry, I just found out that we need an additional conversion to `str`, otherwise the ES tokenizer is complaining when trying to store the data :/

```python
role_files[str(Path(dir.path).relative_to(ROLES_DIRECTORY))] = {
```

Or we do the str conversion later on when writing the data to ES, but doing it directly here feels a little easier.

Reply:
I have converted this here as it is also needed for tests. Conftest.py has the role names as strings so to simplify things I convert path to string in scraper